sql/src/main/java/io/crate/executor/transport/TransportSchemaUpdateAction.java

```diff
@@ -0,0 +1,195 @@
+                                                           TimeValue timeout) {
+        CompletableFuture<SchemaUpdateResponse> future = new CompletableFuture<>();
+        clusterService.submitStateUpdateTask("update-template-mapping", new ClusterStateUpdateTask(Priority.HIGH) {
+            @Override
+            public ClusterState execute(ClusterState currentState) throws Exception {
+                return updateTemplate(xContentRegistry, currentState, indexName, mappingSource);
+            }
+
+            @Override
+            public TimeValue timeout() {
+                return timeout;
+            }
+
+            @Override
+            public void onFailure(String source, Exception e) {
+                future.completeExceptionally(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                future.complete(new SchemaUpdateResponse(true));
+            }
+        });
+        return future;
+    }
+
+    private static ClusterState updateTemplate(NamedXContentRegistry xContentRegistry,
+                                               ClusterState currentState,
+                                               String indexName,
+                                               String mappingSource) throws Exception {
+        PartitionName partitionName = PartitionName.fromIndexOrTemplate(indexName);
+        TableIdent tableIdent = partitionName.tableIdent();
+        String templateName = PartitionName.templateName(tableIdent.schema(), tableIdent.name());
+        IndexTemplateMetaData template = currentState.metaData().templates().get(templateName);
+        if (template == null) {
+            throw new ResourceNotFoundException("Template \"" + templateName + "\" for partitioned table is missing");
+        }
+        XContentParser parser = JsonXContent.jsonXContent.createParser(NamedXContentRegistry.EMPTY, mappingSource);
+        Map<String, Object> newMapping = parser.map();
+
+        IndexTemplateMetaData.Builder templateBuilder = new IndexTemplateMetaData.Builder(template);
+        for (ObjectObjectCursor<String, CompressedXContent> cursor : template.mappings()) {
+            Map<String, Object> source = parseMapping(xContentRegistry, cursor.value.toString());
+            XContentHelper.update(source, newMapping, true);
+            try (XContentBuilder xContentBuilder = JsonXContent.contentBuilder()) {
+                templateBuilder.putMapping(cursor.key, xContentBuilder.map(source).string());
+            }
+        }
+
+        MetaData.Builder builder = MetaData.builder(currentState.metaData()).put(templateBuilder);
+        return ClusterState.builder(currentState).metaData(builder).build();
+    }
+
+    @Override
+    protected ClusterBlockException checkBlock(SchemaUpdateRequest request, ClusterState state) {
+        return null;
```

Comment:
hm, don't we need to check for a metadata write block? if not elaborate as a note why not.

thinking this further, this could even block on sibling partitions?